packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
-      assert(false, "listening with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
+    it("should listen to an event", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const fakeFn = sinon.fake();
 
-  it.skip("should listen to an event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
-    const handlerFunction = async message => {};
+      const handlerFunction = message => {
+        const { data } = message;
+        const { args } = data;
+        fakeFn();
+        subscription.removeListener("ValueChanged", handlerFunction);
+      };
+
+      subscription.on("ValueChanged", handlerFunction);
+
+      simpleStorage.setValue("hello world");
+
+      await mineBlocks(simpleStorage.getProvider(), 10);
+
+      expect(fakeFn.called).to.be.true;
+    });
 
-    subscription.on("ValueChanged", handlerFunction);
+    it("should remove listener after timeout", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const eventFn = sinon.fake();
+      const errorFn = sinon.fake();
+
+      const foreverCallback = async message => {
+        eventFn();
+      };
+
+      await subscription.on("ValueChanged", foreverCallback);
+
+      await subscription.on("error", err => {
+        errorFn();
+      });
+
+      simpleStorage.setValue("hello world");
+
+      await mineBlocks(simpleStorage.getProvider(), 10);
+
+      expect(eventFn.called).to.be.true;
+
+      // TODO: Use fake timer
+      await new Promise(resolve => setTimeout(resolve, 3000));
+
+      expect(
+        subscription.listenerCount("ValueChanged"),
+        "listener should be removed after timeout"
+      ).to.be.equals(0);
+
+      expect(errorFn.called).to.be.true;
+    });
+
+    it("should manage many listeners", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+
+      const listener1 = message => {};
+      const listener2 = message => {};
+      const listener3 = message => {};
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(0);
+
+      subscription.on("ValueChanged", listener1);
+      subscription.on("ValueChanged", listener2);
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(2);
+
+      subscription.off("ValueChanged", listener2);
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(1);
+
+      subscription.on("ValueChanged", listener3);
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(2);
+
+      subscription.removeAllListeners();
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(0);
+    });
   });
 
+  // Send method interface: Contract.send(tx: msg, bool: free) => Subscription
+  // On free send how know if identity-contract should be used?
```

Comment:
Someone could be doing free sends without the meta-transaction's tx costs necessarily being paid out of their identity contract's tokens. Even without identity contracts, a decentralized network of relayers or a centralized relayer might decide to pay the costs for certain meta-transactions that have been broadcast.

Reply:
But there may be more to your question that I'm not getting yet. Could you elaborate on your question?

Reply:
I'll also note that free sends might also happen for higher levels of abstraction than just `Contract.send`. For instance, I might do `erc721Contract.transferFrom(transferFromFunctionParams, free: true)`

Reply:
I'll append your comments to the [related issue](https://github.com/tasitlabs/TasitSDK/issues/88), alright?